homework-solutions/ts/hw-14/task-2.ts

```diff
@@ -0,0 +1,22 @@
+//    Напишите функцию, реализующую метод массивов map(сам мэп юзать нельзя, надо написать кастомный:). Функция принимеют на вход массив и колбэк. Используйте дженерик типы. 
+//    Затипизировать надо саму функцию и коллбэк.
+//    Создать реализацию функции map, принимающую массив чисел 1-5, возвращающую новый массив, 
+//    где каждый каждый элемент - это элемент исходного массива умноженный на его индекс
+//    Пример:
+//    map([1,2,3,4,5], callback) => [0,2,6,12,20]
+
+type MapCallback<T> = (element: T, index: number, array: T[]) => T;
+
+function map<T>(array: T[], callback: MapCallback<T>): T[] {
+    let arr: Array<T> = [];
+    for(let i=0; i < array.length; i++) {
```

Comment:
В результате использования `.map()` можно получить совершенно другой тип, возможно мы захотим привести к значения строке или создать объект
```suggestion
type MapCallback<T, U> = (element: T, index: number, array: T[]) => U;

function map<T, U>(array: T[], callback: MapCallback<T, U>): U[] {
    const arr: Array<U> = [];
    for(let i=0; i < array.length; i++) {
```